src/lib/storage/StorageService.ts

```diff
@@ -0,0 +1,15 @@
+import { BaseRecord } from './BaseRecord';
+
+export interface StorageService<T extends BaseRecord> {
+  save(record: T): Promise<void>;
+
+  update(record: T): Promise<void>;
+
+  delete(record: T): Promise<void>;
+
+  find<T>(typeClass: { new (...args: any[]): T }, id: string, type: string): Promise<T>;
+
+  findAll<T>(typeClass: { new (...args: any[]): T }, type: string): Promise<T[]>;
+
+  findByQuery<T>(typeClass: { new (...args: any[]): T }, type: string, query: {}): Promise<T[]>;
```

Comment:
I've been thinking how to get rid of this `typeClass` and `type` attribute and use just OOP mechanics with static factory method for example, but I couldn't figure anything out 🤷‍♂

Reply:
We can probably get rid of type there because typeClass is actually having a static field of type anyway ;)

`typeClass: { new (...args: any[]): T; type: string }`

The end user however wont deal with it directly anyway since they're going to access everything through the repository and the signature there for find* operations doesn't require any type information. It's supplied while invoking the constructor...